From 439c376d36de75d7560e5a5efa8c00811b365bf0 Mon Sep 17 00:00:00 2001
From: feng wang <feng.a.wang@intel.com>
Date: Mon, 10 Nov 2014 18:58:41 +0800
Subject: [PATCH 5/8] USB IPV6 Mac address causes wrong router IPV6
 configuration

When USB CDC-NCM driver is initialized, a MAC address is defines for this Ethernet interface.
In IPV6 stack, any Ethernet interface creates automatically a link/local address based on MAC.
Even if UE receives a new link/local from Network, IPV6 stack generates or uses automatically
the MAC address link/local insetad of the link/local coming from Network.
This could leads to wrong IPV6 Address assignement and failure in some test cases,
And it could confuse the network because several UE will do a router sollicitation
using same link/local (because the MAC address doesn't change for this cdc interface.

The fix consist to either reset the auto_linklocal (but not supported in theis kernel) or
remove or undefined the MAC address. To do so, we define the MAC length address to 1 like
many other network device driver does.

Change-Id: I52ee3deda05d6673217b1b49d8b81987e5837f56
Signed-off-by: feng wang <feng.a.wang@intel.com>
Reviewed-by: Yang, Fei <fei.yang@intel.com>
Reviewed-by: Tang, Richard <richard.tang@intel.com>
Tested-by: Wang, Ran <ran.wang@intel.com>
Signed-off-by: Li Aixiong <aixiong.li@intel.com>
---
 drivers/net/usb/cdc_ncm.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/net/usb/cdc_ncm.c b/drivers/net/usb/cdc_ncm.c
index 9d54339..2ada082 100644
--- a/drivers/net/usb/cdc_ncm.c
+++ b/drivers/net/usb/cdc_ncm.c
@@ -580,6 +580,13 @@ static int cdc_ncm_bind(struct usbnet *dev, struct usb_interface *intf)
 	return ret;
 }
 
+static int cdc_ncm_remote_mac_bind(struct usbnet *dev,
+					struct usb_interface *intf)
+{
+	dev->net->addr_len = 1;
+	return cdc_ncm_bind(dev, intf);
+}
+
 static void cdc_ncm_align_tail(struct sk_buff *skb, size_t modulus, size_t remainder, size_t max)
 {
 	size_t align = ALIGN(skb->len, modulus) - skb->len + remainder;
@@ -1143,6 +1150,18 @@ static const struct driver_info wwan_noarp_info = {
 	.tx_fixup = cdc_ncm_tx_fixup,
 };
 
+static const struct driver_info cdc_ncm_remote_mac_info = {
+	.description = "CDC NCM",
+	.flags = FLAG_POINTTOPOINT | FLAG_NO_SETINT | FLAG_MULTI_PACKET,
+	.bind = cdc_ncm_remote_mac_bind,
+	.unbind = cdc_ncm_unbind,
+	.check_connect = cdc_ncm_check_connect,
+	.manage_power = usbnet_manage_power,
+	.status = cdc_ncm_status,
+	.rx_fixup = cdc_ncm_rx_fixup,
+	.tx_fixup = cdc_ncm_tx_fixup,
+};
+
 static const struct usb_device_id cdc_devs[] = {
 	/* Ericsson MBM devices like F5521gw */
 	{ .match_flags = USB_DEVICE_ID_MATCH_INT_INFO
@@ -1188,6 +1207,11 @@ static const struct usb_device_id cdc_devs[] = {
 		USB_CDC_SUBCLASS_NCM, USB_CDC_PROTO_NONE),
 	  .driver_info = (unsigned long)&wwan_noarp_info,
 	},
+	{ USB_DEVICE_AND_INTERFACE_INFO(0x1519, 0x0452,
+		USB_CLASS_COMM,
+		USB_CDC_SUBCLASS_NCM, USB_CDC_PROTO_NONE),
+	  .driver_info = (unsigned long)&cdc_ncm_remote_mac_info,
+	},
 
 	/* Generic CDC-NCM devices */
 	{ USB_INTERFACE_INFO(USB_CLASS_COMM,
-- 
1.9.1

